### PR TITLE
Rework randomness state machine & network loop to use watermarks

### PR DESCRIPTION
**Pull request by aschran**
## Description 

We don't need to track individual pending rounds, everything can operate more simply with tracking highest completed round (as reported by checkpoint executor) and highest requested round (as reported by consensus handler).

This change resolves a race bug when checkpoint execution is ahead of consensus.

## Test plan 

Verified with unit tests. Manual test in synthetic environment.

### CODE REVIEW

File: crates/sui-core/src/epoch/randomness.rs [hunk 1]
```diff
@@ -278,13 +278,15 @@ impl RandomnessManager {
...
+        if highest_completed_round + 1 < rm.next_randomness_round {
             info!(
-                "random beacon: resuming generation for randomness round {}",
-                round.0
+                "random beacon: resuming generation for randomness rounds from {} to {}",
+                highest_completed_round + 1,
+                rm.next_randomness_round - 1,
             );
-            network_handle.send_partial_signatures(committee.epoch(), round);
+            for r in highest_completed_round.0 + 1..rm.next_randomness_round.0 {
```
Comment (benr-ml) [inline crates/sui-core/src/epoch/randomness.rs:287]: if the gap is large, this can flood the network, no?

Comment (aschran) [reply crates/sui-core/src/epoch/randomness.rs:287]: No because of the concurrency limit in RandomnessNetworkLoop

Comment (mystenmark) [inline crates/sui-core/src/epoch/randomness.rs:287]: this is most confusing syntax in all of rust 😠

Comment (aschran) [reply crates/sui-core/src/epoch/randomness.rs:287]: I tried to implement the `Step` trait on `RandomnessRound` but the compiler complained it's unstable :(

**Review by mystenmark — APPROVED** [review]

**Pull request merged**